addons/persistence/mongodb-persistence-addon/src/main/java/org/kie/kogito/mongodb/PersistProcessInstances.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright 2020 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.kie.kogito.mongodb;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+
+import org.jbpm.marshalling.impl.ProcessInstanceDocument;
+import org.kie.kogito.mongodb.model.ProcessInstanceModel;
+import org.kie.kogito.mongodb.utils.CommonUtils;
+import org.kie.kogito.process.MutableProcessInstances;
+import org.kie.kogito.process.ProcessInstance;
+import org.kie.kogito.process.ProcessInstanceDuplicatedException;
+import org.kie.kogito.process.impl.AbstractProcessInstance;
+import org.kie.kogito.process.impl.marshalling.ProcessInstanceMarshaller;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.mongodb.client.FindIterable;
+import com.mongodb.client.MongoClient;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoCursor;
+import com.mongodb.client.model.Filters;
+
+@SuppressWarnings({ "rawtypes" })
```

Comment:
Try to use generics for ProcessInstance where possible. I replaced a bunch but I must have missed some others